bycycle/plts/cyclepoints.py

```diff
@@ -0,0 +1,187 @@
+"""Plot extrema and zero-crossings."""
+
+import numpy as np
+
+import matplotlib.pyplot as plt
+
+from neurodsp.plts import plot_time_series
+from neurodsp.plts.utils import savefig
+
+from bycycle.utils import limit_df, limit_sig_times, get_extrema
+
+###################################################################################################
+###################################################################################################
+
+@savefig
+def plot_cyclepoints_df(df, sig, fs, xlim=None, plot_sig=True, plot_extrema=True,
+                        plot_zerox=True, ax=None, **kwargs):
+    """Plot extrema and/or zero-crossings using a dataframe to define points.
+
+    Parameters
+    ----------
+    df : pandas.DataFrame
+        Dataframe output of :func:`~.compute_features`.
+    sig : 1d array
+        Time series to plot.
+    fs : float
+        Sampling rate, in Hz.
+    xlim : tuple of (float, float), optional, default: None
+        Start and stop times.
+    plot_sig : bool, optional, default: True
+        Plots the raw signal.
+    plot_extrema :  bool, optional, default: True
+        Plots peaks and troughs.
+    plot_zerox :  bool, optional, default: True
+        Plots zero-crossings.
+    ax : matplotlib.Axes, optional, default: None
+        Figure axes upon which to plot.
+    **kwargs
+        Keyword arguments to pass into `plot_time_series`.
+
+    Notes
+    -----
+    Default keyword arguments include:
+
+    - ``figsize``: tuple of (float, float), default: (15, 3)
+    - ``xlabel``: str, default: 'Time (s)'
+    - ``ylabel``: str, default: 'Voltage (uV)
+
+    """
+
+    # Set default kwargs
+    figsize = kwargs.pop('figsize', (15, 3))
+    xlabel = kwargs.pop('xlabel', 'Time (s)')
+    ylabel = kwargs.pop('ylabel', 'Voltage (uV)')
+
+    # Set times and limits
+    times = np.arange(0, len(sig) / fs, 1 / fs)
+    xlim = (times[0], times[-1]) if xlim is None else xlim
+
+    # Determine extrema/zero-crossing times and signals
+    center_e, side_e = get_extrema(df)
+
+    df = limit_df(df, fs, xlim)
+    sig, times = limit_sig_times(sig, times, xlim)
+
+    # Extend plotting based on given arguments
+    x_values = []
+    y_values = []
+    colors = ['k']
+
+    if plot_extrema:
```

Comment:
this function feels like it could do less work, given `plot_cyclepoints_array` exists. 

Why not:
```
peaks, troughs, rises, falls = None, None, None, None
if plot_extrema:
    # extract peaks & troughs from df and assign to variables
if plot_zerox:
    # extract rises & decays from df and assign to variables

plot_cyclepoints_array(sig, fs, peaks, troughs, rises, decays, **kwargs)
```

The only thing that might need to be updated / coordinated is the make consistent if & how `plot_cyclepoints_array` handles not plotting a signal (which is an option from the df func), but I definitely think we want to reduce what is otherwise copied code here.

Reply:
This function only uses the df from `compute_features` to determine cyclepoints. If someone first uses `compute_features`, cyclepoint arrays are not accessible. This is where I'd argue where a class structure may come in handy to carry around meta-data that may or may not be needed later. I've avoid this for now unless more cases like this come up.

To avoid recomputing the cyclepoints, the current implementation uses the df to determine cyclepoints instead. This is more efficient, computationally, than recomputing with `find_extrema` and `find_zerox`. If this function were to recompute these arrays, it could be dropped altogether, since the df will no longer be needed.

I should add the option for plotting the signal in `plot_cyclepoints_array`. The reason this is an option in the df functions, is for a case when `plot_bursts` is used first, and then cyclepoints points are overlayed. This was need to for the summary plotting function.

If copied code is a concern, I could move it to a private func that is used by both functions?

Reply:
I replied but then I realized I misunderstood you point and deleted. This makes since to reduce code. :slightly_smiling_face: